File: programmers/[1차] 다트 게임/solution.cpp
#include <string>
#include <stack>
#include <cmath>

using namespace std;

int solution(string dr)
{
    int sum = 0;
    stack<int> s;
    int temp = 0;
    int temp2 = 0;

    for (int i = 0; i < dr.size(); i++)
    {
        if (dr[i] == '1')
        {
            s.push(dr[i] - '0');

            if (dr[i + 1] == '0')
            {
                s.pop();
                s.push(10);
                i++;
            }

            if (dr[i + 1] == 'D')
            {
                temp = s.top();
                s.pop();
                s.push(pow(temp, 2));
            }

            else if (dr[i + 1] == 'T')
            {
                temp = s.top();
                s.pop();
                s.push(pow(temp, 3));
            }
        }
        else
        {
            s.push(dr[i] - '0');

            if (dr[i + 1] == 'D')
            {
                temp = s.top();
                s.pop();
                s.push(pow(temp, 2));
            }
            else if (dr[i + 1] == 'T')
            {
                temp = s.top();
                s.pop();
                s.push(pow(temp, 3));
            }
        }
        i++;

        if (dr[i + 1] == '*' || dr[i + 1] == '#')
        {
            if (dr[i + 1] == '*')
            {
                if (s.size() == 1)
                {
                    temp = s.top();
                    s.pop();
                    s.push(temp * 2);
                }
                else
                {
                    temp2 = s.top();
                    s.pop();
                    temp = s.top();
                    s.pop();
                    temp *= 2;
                    s.push(temp);
                    temp2 *= 2;
                    s.push(temp2);
                }
            }
            else
            {
                temp = s.top();
                s.pop();
                s.push(-temp);
            }
            i++;
        }
    }

    while (!s.empty())
    {
        sum += s.top();
        s.pop();
    }

    return sum;
}

File: programmers/[1차] 다트 게임/Solution.java
import java.util.*;

class Solution {
    public int solution(String dartResult) {
        dartResult = dartResult.replace("10", "x");
        int answer = 0;
        
        ArrayList<Integer> scoreList = new ArrayList<>();
        
        int i = 0;
        while (i < dartResult.length()) {
            int scoreIndex = scoreList.size() - 1;
            
            switch (dartResult.charAt(i)) {
                case 'S':
                    scoreList.set(scoreIndex, (int) Math.pow(scoreList.get(scoreIndex), 1));
                    break;
                case 'D':
                    scoreList.set(scoreIndex, (int) Math.pow(scoreList.get(scoreIndex), 2));
                    break;
                case 'T':
                    scoreList.set(scoreIndex, (int) Math.pow(scoreList.get(scoreIndex), 3));
                    break;
                case '*':
                    if (scoreIndex < 1) {
                        scoreList.set(scoreIndex, scoreList.get(scoreIndex) * 2);
                    } else {
                        scoreList.set(scoreIndex, scoreList.get(scoreIndex) * 2);
                        scoreList.set(scoreIndex - 1, scoreList.get(scoreIndex - 1) * 2);
                    }
                    break;
                case '#':
                    scoreList.set(scoreIndex, scoreList.get(scoreIndex) * -1);
                    break;
                case 'x':
                    scoreList.add(10);
                    break;
                default:
                    scoreList.add(Character.getNumericValue(dartResult.charAt(i)));
            }
            
            i++;
        }
        
        return scoreList.stream().mapToInt(Integer::valueOf).sum();
    }
}